nupic/research/frameworks/vernon/core_mixins/step_based_logging.py

```diff
@@ -0,0 +1,153 @@
+        super().setup_experiment(config)
+        self.current_timestep = 0
+        self.log_timestep_freq = config.get("log_timestep_freq", 1)
+
+    def run_epoch(self):
+        timestep_begin = self.current_timestep
+        ret = super().run_epoch()
+        ret.update(
+            timestep_begin=timestep_begin,
+            timestep_end=self.current_timestep,
+        )
+        return ret
+
+    def post_batch(self, **kwargs):
+        super().post_batch(**kwargs)
+        self.current_timestep += 1
+
+    def should_log_batch(self, train_batch_idx):
+        """
+        Returns true if the current timestep should be logged, either because it's a
+        logged timestep or the final training batch of an epoch.
+
+        This is a utility method, not intended for extensibility.
+        """
+        return (train_batch_idx == self.total_batches - 1) or (
+            self.log_timestep_freq > 0
+            and (self.current_timestep % self.log_timestep_freq) == 0)
+
+    def get_state(self):
+        state = super().get_state()
+        state["current_timestep"] = self.current_timestep
+        return state
+
+    def set_state(self, state):
+        super().set_state(state)
+        if "current_timestep" in state:
+            self.current_timestep = state["current_timestep"]
+
+    @classmethod
+    def get_recorded_timesteps(cls, result, config):
+        """
+        Given an epoch result dict and config, returns a list of timestep numbers
+        that are supposed to be logged for that epoch.
+
+        This is a utility method, not intended for extensibility.
+        """
+        log_timestep_freq = config.get("log_timestep_freq", 1)
+        timestep_end = result["timestep_end"]
+        if log_timestep_freq == 0:
+            ret = [timestep_end - 1]
+        else:
+            # Find first logged timestep in range
+            logged_begin = int(math.ceil(result["timestep_begin"]
+                                         / log_timestep_freq)
+                               * log_timestep_freq)
+
+            ret = list(range(logged_begin, timestep_end, log_timestep_freq))
+
+            last_batch_timestep = timestep_end - 1
+            if last_batch_timestep % log_timestep_freq != 0:
+                ret.append(last_batch_timestep)
+
+        return ret
+
+    @classmethod
+    def expand_result_to_time_series(cls, result, config):
+        """
+        Given a result dict containing data for multiple batches, returns a mapping
+        from timesteps to results. The mapping is stored as a dict so that
+        subclasses and mixins can easily add data to it.
+
+        Result keys are converted from Ray Tune requirements to better names,
+        and the keys are filtered to those that make useful charts.
+
+        :return: defaultdict mapping timesteps to result dicts
+        """
+        result_by_timestep = defaultdict(dict)
+
+        # Assign the epoch result to the appropriate timestep.
+        k_mapping = {
```

Comment:
I see `k_mapping` is defined here and in `ExtraValidation`. Maybe we can make this into a class function (which I believe we've discussed before)?

It would help reduce redundancy, for one. But, more importantly, it leaves `StepBasedLogging` more agnostic to what's being logged at every step. For instance, the `ContinualLearningExperiment` also trains in batches but uses different metrics.